sonarts-core/src/rules/cognitiveComplexityRule.ts

```diff
@@ -0,0 +1,311 @@
+  }
+
+  public static DEFAULT_THRESHOLD = 15;
+
+  public static TARGETED_KINDS = [
+    // ArrowFunctions and FunctionExpressions are included in computation
+    ts.SyntaxKind.FunctionDeclaration,
+    ts.SyntaxKind.MethodDeclaration,
+    ts.SyntaxKind.Constructor,
+    ts.SyntaxKind.GetAccessor,
+    ts.SyntaxKind.SetAccessor,
+  ];
+}
+
+class FunctionWalker extends tslint.RuleWalker {
+  threshold: number;
+
+  constructor(sourceFile: ts.SourceFile, options: tslint.IOptions, threshold: number) {
+    super(sourceFile, options);
+    this.threshold = threshold;
+  }
+
+  public visitNode(node: ts.Node): void {
+    if (is(node, ...Rule.TARGETED_KINDS)) {
+      const body = (node as ts.FunctionLikeDeclarationBase).body as ts.Node;
+      if (body) {
+        const complexity = this.getComplexity(body);
+        if (complexity > this.threshold) {
+          this.addFailureAtNode(
+            FunctionWalker.reportNode(node),
+            FunctionWalker.getMessage(node, complexity, this.threshold),
+          );
+        }
+      }
+    }
+    super.visitNode(node);
+  }
+
+  private getComplexity(node: ts.Node): number {
+    let complexityWalker = new ComplexityWalker(this.getSourceFile(), this.getOptions());
+    complexityWalker.walk(node);
+    return complexityWalker.getComplexity();
+  }
+
+  private static reportNode(node: ts.Node): ts.Node {
+    let reportNode;
+    switch (node.kind) {
+      case ts.SyntaxKind.FunctionDeclaration:
+        reportNode = (node as ts.FunctionDeclaration).name;
+        break;
+      case ts.SyntaxKind.MethodDeclaration:
+        reportNode = (node as ts.MethodDeclaration).name;
+        break;
+      case ts.SyntaxKind.Constructor:
+        reportNode = node.getFirstToken();
+        break;
+      case ts.SyntaxKind.GetAccessor:
+        reportNode = (node as ts.GetAccessorDeclaration).name;
+        break;
+      case ts.SyntaxKind.SetAccessor:
+        reportNode = (node as ts.SetAccessorDeclaration).name;
+        break;
+    }
+    if (!reportNode) {
+      return node;
+    }
+    return reportNode;
+  }
+
+  private static getMessage(node: ts.Node, complexity: number, threshold: number): string {
+    const functionName = FunctionWalker.getFunctionName(node);
+    return `Refactor this ${functionName} to reduce its Cognitive Complexity from ${complexity} to the ${threshold} allowed.`;
+  }
+
+  private static getFunctionName(node: ts.Node): string {
+    switch (node.kind) {
+      case ts.SyntaxKind.MethodDeclaration:
+        return "method";
+      case ts.SyntaxKind.Constructor:
+        return "constructor";
+      case ts.SyntaxKind.GetAccessor:
+      case ts.SyntaxKind.SetAccessor:
+        return "accessor";
+      default:
+        return "function";
+    }
+  }
+}
+
+class ComplexityWalker extends tslint.RuleWalker {
+  private complexity = 0;
+  private nesting = 0;
+  private logicalOperationsToIgnore: ts.Node[] = [];
```

Comment:
@m-g-sonar can we use a more precise type here?

Reply:
Yup, we only add `ts.BinaryExpression` here.